feat: introduce GardenEditor with theming

austingreendev opened the pull request:
## Description

This PR contains some initial steps toward theming CKEditor to align with Garden standards.

## Details

The first focus for this PR is to theme some of the basic typography and toolbar colors/layouts.

In later PRs we will be creating a `GardenEditor` which will provide these themed values for consumers.

## Theming covered in this PR

- Update select chevron icon
- 2px between toolbar items
- Cursor changes to a pointer on hover
- Bold treatment uses semibold
- Update localized strings
- Inline code border radius
- Tooltip border radius
- Toolbar separator color
- Heading sizing
- Remove mobile typearound buttons

## Not covered

Some areas that aren't implemented as of yet:
- Dropdown styling
- Link tooltip balloon styling
- Relative spacing between content
- List styling

austingreendev commented on src/stories/GardenEditor.stories.js:
```diff
@@ -5,34 +5,127 @@
+  'Increase indent': `Increase indent (${getEnvKeystrokeText('CTRL+]')})`,
+  'Block quote': `Block quote (${getEnvKeystrokeText('CTRL+SHIFT+9')})`,
+  Code: `Code (${getEnvKeystrokeText('CTRL+SHIFT+5')})`,
+  'Horizontal line': `Horizontal rule (${getEnvKeystrokeText('CTRL+SHIFT+L')})`
```
These are key-value pairs for localization. CKEditor refers to these as `line` while we use `rule`. This makes it consistent with Agent workspace.

jzempel commented on src/stories/theme-styling.css:
```diff
@@ -0,0 +1,204 @@
+}
+
+.ck.ck-editor__editable .ck-code_selected {
+  background-color: #d8dcde; /* GREY-300 */
```
@ginnywood @m-lai why is it desirable to darken the background color on `Code` selection? This doesn't seem consistent with the preformatted text block.

ginnywood replied:
That is a good point. Let’s remove it.

austingreendev replied:
Since `ck-code_selected` is a part of the `code` plugin we still have to override the original value. I've updated this to match the default `grey-200` state.

jzempel reviewed (commented):
@austingreendev can you break `theme-styling.css` up into individual component CSS files? It seems like that would be a nice win for code maintenance.

austingreendev commented on src/stories/theme-styling.css:
```diff
@@ -0,0 +1,204 @@
+  margin-block-end: 16px;
+}
+
+/* LISTS */
```
Does this relate to overall line-height within the editor? I didn't see any line-heights defined within https://github.com/zendeskgarden/react-components/blob/main/packages/typography/src/styled/StyledList.ts

We also modify the relative margins to handle nesting and other text content.

austingreendev commented on src/stories/theme-styling.css:
```diff
@@ -0,0 +1,204 @@
+  margin: var(--ck-spacing-standard) 0;
+}
+
+/* Blockquote */
```
`calc()` included too much padding by default. This is now updated.

austingreendev commented:
@jzempel this `theme-styling` will eventually not exist and be moved fully within the `src/theme` directory. Breaking this into separate files for now.

jzempel commented on src/stories/initialData.js:
```diff
@@ -0,0 +1,9 @@
+ */
+
+// eslint-disable-next-line max-len
+export const INITIAL_DATA = `<h2>CKEditor5</h2><hr><h3>Basic styles</h3><p>Text can have <strong>bold</strong>, <i>italic</i>, and <u>underline</u> content. <code>Inline code</code> is also available. <a href="https://www.google.com">Links work as well.</a></p><ul><li>Lists</li><li>Are<ol><li>Nestable</li><li>And can change</li><li>Types</li></ol></li></ul><h3>Block elements</h3><blockquote><p>Bl ...
```
Please use "CKEditor" throughout. Primarily for maintenance, as someday the version will change. (Secondarily, CKSource writes it as "CKEditor<space>5".)

jzempel commented on src/stories/theme/_list.css:
```diff
@@ -0,0 +1,33 @@
+.ck-content ul {
+  margin: 16px 0px 16px 20px;
```
Garden's list `margin-left` is 24px. Garden's list `<li>`s are 24px in height – 20px line height + 4px padding. I realize you are locked into a semantic ul/ol -> li structure here, but options should be explored (`::before`/`::after`?) for getting the spacing right.

austingreendev replied:
Almost all of the content needs some work when it comes to margin and relative spacing. Might be worthwhile to handle that all together.

jzempel commented on src/stories/theme/_list.css:
```diff
@@ -0,0 +1,33 @@
+  margin: 16px 0px 16px 20px;
+  padding: 0;
+  list-style-position: outside;
+  list-style-type: disc;
```
Ideally, there would be styling that corresponds with the nested form seen under https://garden.zendesk.com/components/lists#nesting. Following Google docs, this is `1. - a. - i.` (repeating) for ordered and `disc - circle - square` (repeating) for unordered.

austingreendev replied:
We are unable to enforce nested list sequencing as that can be controlled via a separate plugin that allows consumers to choose them directly. [Here is an example of those dropdowns in action.](https://ckeditor.com/ckeditor-5/demo/#document-paged)

This is also business logic, correct? Guide currently enforces these values from what is in the Copenhagen theme, which doesn't align with our values.

jzempel replied:
In the case where ordering cannot be automated, it is still important to capture and expose the Garden opinion on this matter. In this case, there should be a plan for some kind of "implementation notes" documentation that points to the plugin and sets expectations for the default ordering to be used within Zendesk. Whether that is exposed within the repo's README, within internal design/dev documentation, or some other form is a topic for team conversation that should be fleshed out and determined.

austingreendev replied:
For sure. I'm taking some additional notes in the audit doc so we can return to this.

jzempel commented on src/stories/GardenEditor.stories.js:
```diff
@@ -5,34 +5,127 @@
+          'bold',
+          'italic',
+          'underline',
+          '|',
```
Can separators be handled responsively? It feels strange that this appears before the overflow menu when the editor shrinks down to "small mobile" size.

austingreendev replied:
Weird, this is being removed as I would expect in mobile for me 
![image](https://user-images.githubusercontent.com/4030377/104218969-5947c980-53f2-11eb-9b4b-50564de87ea1.png)

jzempel replied:
I can see the vertical separator before the overflow button in the screen shot you provided ☝️

austingreendev replied:
I think that is a separate `ck-separator` that is injected by the toolbar. It's not controlled by the toolbar config.

jzempel replied:
Not a super high priority, but can this be dealt with via CSS? We should make this thing bend to our styling will wherever possible.

austingreendev replied:
It looks like it's appended to the DOM and doesn't include any targetable classes. I'll make a not to take a look at this along with the other toolbar customizations.

jzempel commented:
Garden's `stylelint-config` should be added to keep everything nice and clean.

ginnywood reviewed (changes requested):
I know this PR addresses only the items marked “needs review” in our doc, so I’ll comment on those and leave the items that are in progress or not started.

These look good so far:
1. Select chevron is 16-chevron-down icon
2. 2px between toolbar items
3. Cursor changes to a pointer on hover
4. Bold treatment uses semibold
5. Strings are looking good to me. @ctancinco can you review as well?
6. Inline code border radius is 2px
7. Tooltip: corner radius should be 2px instead of 4px. Height (without arrow) should be 20px instead of 18px. I’m looking into whether it should touch the button.

These need to be fixed:
1. Lines between toolbar sections should be grey-200 and 24px tall or 60% of toolbar height (or the rem equivalent)
2. Type sizes: heading 1 should be our XXL, heading 2 should be our XL, heading 3 should be our L
3. Lists should match Garden lists line height, indentation, and nested bullet types

austingreendev commented:
@ginnywood the typography was getting shifted from `h1` to `h2`. The heads should now be the correct size.

austingreendev commented:
@jzempel @ginnywood What do you think of having all spacing/sizing concerns handled in a separate PR? I think there are enough questions around relative spacing that we can solve them all at once, including the `list` spacing.

This might be a good stopping point for the initial PR so we can dig into dropdowns and more advanced styling in the next one.

jzempel commented:
I'm all for kicking the can down the road. My only concern is wasting review time by failing to address this feedback in the future – prior to shipping a version. In other words, don't make reviewers manage the "work to be done". @ginnywood what do you think?

@austingreendev what about https://github.com/zendeskgarden/ckeditor/pull/3#issuecomment-757013383

ginnywood commented:
I think that’s a logical way to separate the PRs. @austingreendev could you make sure you update the description of this PR with a list of what you’re addressing?

It is true that we’ll need a way to track what’s being done in the future. What about this: if you’re punting on one of the things I mentioned, change the status in the doc back to “in progress?”

austingreendev commented:
@jzempel I've included the Garden stylelint config, added it to our standard pre-commit and lint runs, and corrected any failures.

I've updated the description to cover what is already marked as "needs review" in the audit document. The content in that document is continuing to change so let's just review what's in the PR description.

jzempel reviewed (approved):
works for me 🙏

ginnywood reviewed (approved):
Everything scoped for this PR is looking good. Thanks @austingreendev!
